doc/sphinx/proof-engine/tactics.rst

```diff
@@ -105,6 +105,227 @@ bindings_list`` where ``bindings_list`` may be of two different forms:
   .. exn:: Not the right number of missing arguments.
      :undocumented:
 
+.. _intropatterns:
+
+Intro patterns
+~~~~~~~~~~~~~~
+
+Intro patterns let you specify the name to assign to variables and hypotheses
+introduced by tactics.  They also let you split an introduced hypothesis into
+multiple hypotheses or subgoals.  Common tactics that accept intro patterns
+include :tacn:`intros`, :tacn:`destruct` and :tacn:`apply`.
+
+.. productionlist::        coq
+   simple_intropattern     : `simple_intropattern_closed`
+                           : % `operconstr` ... % `operconstr`
+   simple_intropattern_closed : `or_and_intropattern`
+                              : `equality_intropattern`
+                              : _
+                              : `naming_intropattern`
+   or_and_intropattern     : [ `intropatterns` | ... | `intropatterns` ]
+                           : ()
+                           : ( `simple_intropattern` , ... , `simple_intropattern` )
+                           : ( `simple_intropattern` & ... & `simple_intropattern` )
+   equality_intropattern   : ->
+                           : <-
+                           : [= `intropatterns` ]
+   intropatterns           : `nonsimple_intropattern` ...
+                           : <empty>
+   ne_intropatterns        : `nonsimple_intropattern` ...
+   nonsimple_intropattern  : `simple_intropattern`
+                           : *
+                           : **
+   naming_intropattern     : `pattern_ident`
+                           : ?
+                           : `ident`
+   pattern_ident           : ? `ident`
+
+Use these elementary patterns to specify a name:
+
+* :n:`@ident` - use the specified name
+* :n:`?` - let Coq choose a name
+* :n:`? @ident` - generate a name that begins with :n:`@ident`
+* :n:`_` - discards the matched part (unless it is required for another
+  hypothesis)
+* if a disjunction pattern omits a name, such as :g:`[|H2]`, Coq will choose a name
+
+The most common splitting patterns are:
+
+* split a hypothesis in the form :g:`A /\ B` into two
+  hypotheses :g:`H1: A` and :g:`H2: B` using the pattern :g:`[H1 H2]`
+* split a hypothesis in the form :g:`A \/ B` into two
+  subgoals using the pattern :g:`[H1|H2]`.  The first subgoal will have the hypothesis
+  :g:`H1: A` and the second subgoal will have the hypothesis :g:`H2: B`.
+
+**:ref:`Example <_intropattern_conj_ex>` doesn't work for linking to the example**
+
+**Sphinx says: "WARNING: undefined label: _intropattern_conj_ex (if the link has no caption
+the label must precede a section header)"**
+
+Splitting patterns can have more than 2 names, for example :g:`[H1 H2 H3]`
+matches :g:`A /\ B /\ C`.  Patterns can be nested:
```

Comment:
That's incorrect. Because of the implicit parenthesizing the actual pattern used to destruct this is `[H1 [H2 H3]]`, alternatively `(H1, (H2, H3))` or the much nicer (and thus recommended) `(H1 & H2 & H3)` that is made just for that purpose.

Reply:
The & form is recommended over the , form?  Both the [ ] and the , form are not recommended (maybe deprecated someday?)

Reply:
The point of the form with the comma is to look like a pair when destructuring something that can be understood as a pair. However; the intro-pattern with square brackets is more generic. The following cannot be done with any other intro pattern:

```
Inductive foo := | bla : nat -> nat -> foo | bar : nat -> foo.
Goal foo -> True.
intros [n1 n2|n3].
```